### PR TITLE
Add new functional tests suite

### PR DESCRIPTION
**Pull request by mapedraza**
This PR includes a new "functional" test suite. It should:

**Things to have into account in this suite**:

- [x] Be more verbose (including diffs)
- [x] More legible and easy to build
- [x] Make tests independents between them

### CODE REVIEW

File: test/functional/config-test.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        type: 'memory'
+    },
+    types: {},
+    service: 'howtoservice',
+    subservice: '/howto',
+    providerUrl: 'http://localhost:4041',
+    deviceRegistrationDuration: 'P1M',
+    defaultType: 'Thing',
+    defaultResource: '/iot/json',
+    compressTimestamp: true
```
Comment (mrutid) [inline test/functional/config-test.js:68]: why it appears again? in any case should be false.

File: test/functional/functional-tests.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    });
+
+    afterEach(function (done) {
+        nock.cleanAll();
+
+        async.series([iotAgentLib.clearAll, iotaJson.stop], done);
+    });
+
+    describe('Basic group provision without attributes', function () {
+        // Provision
```
Comment (mrutid) [inline test/functional/functional-tests.js:57]: unnecessary comment

File: test/functional/functional-tests.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        const provision = {
+            url: 'http://localhost:' + config.iota.server.port + '/iot/services',
+            method: 'POST',
+            json: {
+                services: [
+                    {
+                        resource: '/iot/json',
+                        apikey: '123456',
+                        entity_type: 'TheLightType',
+                        trust: '8970A9078A803H3BL98PINEQRW8342HBAMS',
```
Comment (mrutid) [inline test/functional/functional-tests.js:67]: I think we ar not testing trusttoken

**Review by mrutid — APPROVED** [review]
Minor.
The structure looks good to me.

File: test/functional/config-test.js (hunk 1 above)
Comment (mapedraza) [reply test/functional/config-test.js:68]: It is inherited from previous unit tests config file. I will remove here and open a new PR to remove it from the rest of config files

File: test/functional/functional-tests.js (hunk 3 above)
Comment (mapedraza) [reply test/functional/functional-tests.js]: Fixed in https://github.com/telefonicaid/iotagent-json/pull/769/commits/0a8d8c21943bdc2c8fa8564b3e109ae17f08b94a

File: test/functional/functional-tests.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        it('should send its value to the Context Broker', function (done) {
+            request(measure, function (error, result, body) {
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('Basic group provision with attributes', function () {
```
Comment (manucarrace) [inline test/functional/functional-tests.js:129]: I propose to refer to the attributes as "active attributes"

File: test/functional/functional-tests.js (hunk 2 above)
Comment (mapedraza) [reply test/functional/functional-tests.js]: Fixed in https://github.com/telefonicaid/iotagent-json/pull/769/commits/0a8d8c21943bdc2c8fa8564b3e109ae17f08b94a

**Comment by mapedraza** [comment]
After committing https://github.com/telefonicaid/iotagent-json/pull/769/commits/ce2e261bc92a794d9ed1f9ff0727a7b31c1088e0 tests are more verbose, you can check how an expectation that differs from the CB payload looks like:
https://github.com/telefonicaid/iotagent-json/actions/runs/6589901511/job/17905362907?pr=769#step:5:287

It is made to intentionally make tests fail for demonstration purposes. 

CC: @manucarrace @mrutid

**Comment by mapedraza** [comment]
It also detects problems when sending the request (southbound) or when provisioning wrongly:
image

File: package.json [hunk 5]
```diff
@@ -29,12 +29,15 @@
...
     "test": "nyc --reporter=text mocha --recursive 'test/**/*.js' --reporter spec --timeout 5000 --ui bdd --exit  --color true",
+    "test:functional": "nyc --reporter=text mocha --recursive 'test/functional/*.js' --reporter spec --timeout 5000 --ui bdd --exit  --color true",
     "test:coverage": "nyc --reporter=lcov mocha -- --recursive 'test/**/*.js' --reporter spec --exit",
     "test:coveralls": "npm run test:coverage && cat ./coverage/lcov.info | coveralls && rm -rf ./coverage",
     "test:watch": "npm run test -- -w ./lib",
     "watch": "watch 'npm test && npm run lint' ./lib ./test"
   },
   "devDependencies": {
+    "async-mqtt": "~2.6.3",
+    "chai": "^4.3.10",
```
Comment (fgalan) [inline package.json:40]: Don't use ^. Always ~

;)

File: package.json [hunk 6]
```diff
@@ -29,12 +29,15 @@
     "prettier:text": "prettier 'README.md' 'docs/*.md' 'docs/**/*.md' --no-config --tab-width 4 --print-width 120 --write --prose-wrap always",
     "start": "node ./bin/iotagent-json",
     "test": "nyc --reporter=text mocha --recursive 'test/**/*.js' --reporter spec --timeout 5000 --ui bdd --exit  --color true",
+    "test:functional": "nyc --reporter=text mocha --recursive 'test/functional/*.js' --reporter spec --timeout 5000 --ui bdd --exit  --color true",
```
Comment (fgalan) [inline package.json:32]: A GitAction for this test should be included.

**Comment by mapedraza** [comment]
The test that expects the measures includes metadata with TimeInstant when `timestamp=true` is failing:

```
A - WHEN sending a measure not named TimeInstant through http IT should add the timestamp to the attributes sent to Context Broker
```

**Provision**
```
                    {
                        resource: '/iot/json',
                        apikey: globalEnv.apikey,
                        entity_type: globalEnv.entity_type,
                        timestamp: true,
                        commands: [],
                        lazy: [],
                        attributes: [
                            {
                                object_id: 'a',
                                name: 'attr_a',
                                type: 'Number'
                            }
                        ]
                    }
````

**expectation**
```
{
                    id: globalEnv.entity_name,
                    type: globalEnv.entity_type,
                    attr_a: {
                        value: 23,
                        type: 'Number',
                        metadata: {
                            TimeInstant: {
                                type: 'DateTime',
                                value: '2015-12-14T08:06:01.468Z'
                            }
                        }
                    }
```

File: package.json (hunk 6 above)
Comment (fgalan) [reply package.json:32]: In that case, why don't just keep `test` target and remove `test:functional`?

Not sure if having overlapping targets is a good idea...

File: package.json [hunk 7]
```diff
@@ -63,7 +67,7 @@
     "body-parser": "1.20.0",
     "dateformat": "3.0.3",
     "express": "4.18.1",
-    "iotagent-node-lib": "https://github.com/telefonicaid/iotagent-node-lib.git#master",
+    "iotagent-node-lib": "github:telefonicaid/iotagent-node-lib#task/add-functional-tests",
```
Comment (fgalan) [inline package.json:70]:
```suggestion
    "iotagent-node-lib": "https://github.com/telefonicaid/iotagent-node-lib.git#master",
```

**Comment by fgalan** [comment]
As discussed at Teams, a test/functional/README.md would be a good idea.

**Review by fgalan — APPROVED** [review]
LGTM

**Pull request merged**